src/azure-cli-core/azure/cli/core/cloud.py

```diff
@@ -0,0 +1,181 @@
+#---------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+#---------------------------------------------------------------------------------------------
+
+import os
+from six.moves import configparser
+
+from azure.cli.core._config import GLOBAL_CONFIG_DIR
+
+CUSTOM_CLOUD_CONFIG_FILE = os.path.join(GLOBAL_CONFIG_DIR, 'clouds.config')
+
+class CloudNotRegisteredException(Exception):
+    def __init__(self, cloud_name):
+        super(CloudNotRegisteredException, self).__init__(cloud_name)
+        self.cloud_name = cloud_name
+    def __str__(self):
+        return "The cloud '{}' is not registered.".format(self.cloud_name)
+
+class CloudAlreadyRegisteredException(Exception):
+    def __init__(self, cloud_name):
+        super(CloudAlreadyRegisteredException, self).__init__(cloud_name)
+        self.cloud_name = cloud_name
+    def __str__(self):
+        return "The cloud '{}' is already registered.".format(self.cloud_name)
+
+class CannotUnregisterCloudException(Exception):
+    pass
+
+class CloudEndpointUrl(object): # pylint: disable=too-few-public-methods
+    MANAGEMENT = 'management'
+    RESOURCE_MANAGER = 'resource_manager'
+    SQL_MANAGEMENT = 'sql_management'
+    GALLERY = 'gallery'
+    ACTIVE_DIRECTORY = 'active_directory'
+    ACTIVE_DIRECTORY_RESOURCE_ID = 'active_directory_resource_id'
+    ACTIVE_DIRECTORY_GRAPH_RESOURCE_ID = 'active_directory_graph_resource_id'
+
+class CloudParam(object): # pylint: disable=too-few-public-methods
+    SQL_SERVER_HOSTNAME_SUFFIX = 'sql_server_hostname_suffix'
+    STORAGE_ENDPOINT_SUFFIX = 'storage_endpoint_suffix'
+    KEYVAULT_DNS_SUFFIX = 'keyvault_dns_suffix'
+    AZURE_DATALAKE_STORE_FILE_SYSTEM_ENDPOINT_SUFFIX = 'azure_datalake_store_file_system_endpoint_suffix' #pylint: disable=line-too-long
+    AZURE_DATALAKE_ANALYTICS_CATALOG_AND_JOB_ENDPOINT_SUFFIX = 'azure_datalake_analytics_catalog_and_job_endpoint_suffix' #pylint: disable=line-too-long
+
+class Cloud(object): # pylint: disable=too-few-public-methods
+    """ Represents an Azure Cloud instance """
+
+    def __init__(self, name, endpoints=None, params=None):
+        self.name = name
+        self.endpoints = endpoints
+        self.params = params
+
+AZURE_PUBLIC_CLOUD = Cloud('AzureCloud', endpoints={
```

Comment:
So, in general I recommend that whatever mechanism we plan to use for Azure Stack, which I believe would be external files containing this info somewhere in the user directory, we use for these as well. The reason being that if we have one path for the "common clouds" and a separate path for Azure stack (or the UNcommon clouds) I worry that while we have plenty of operational testing of the common scenarios, there may be bugs in the alternative scenario that would linger and appear surface to arguably our most important customers.

If the default clouds share the same mechanisms with custom clouds, we lessen that risk. I think we could just save this configuration to individual files that we package automatically in the release.